### ISSUE TITLE
Bring back System.IO.Packaging API

### ISSUE THREAD

**Issue by ericstj** [dotnet/corefx#29959]
Much of the functionality was previously subsetted from System.IO.Packaging.  We need to bring this back for .NETCore 3.0.  We also need to decide if it goes in System.IO.Packaging or if we merely provide a means for extending System.IO.Packaging by some other assembly.  We also have to consider the implementation of below API and what's required.  I believe some was directly interacting with ZLIB in the past.  If we want to facilitate that as public API we may need to expose more from System.IO.Compression.

Here's the full set of API:
```c#
namespace System.IO.Packaging {
    public enum CertificateEmbeddingOption {
        InCertificatePart = 0,
        InSignaturePart = 1,
        NotEmbedded = 2,
    }
    public class EncryptedPackageEnvelope : IDisposable {
        public FileAccess FileOpenAccess { get; }
        public PackageProperties PackageProperties { get; }
        public RightsManagementInformation RightsManagementInformation { get; }
        public StorageInfo StorageInfo { get; }
        public void Close();
        public static EncryptedPackageEnvelope Create(Stream envelopeStream, PublishLicense publishLicense, CryptoProvider cryptoProvider);
        public static EncryptedPackageEnvelope Create(string envelopeFileName, PublishLicense publishLicense, CryptoProvider cryptoProvider);
        public static EncryptedPackageEnvelope CreateFromPackage(Stream envelopeStream, Stream packageStream, PublishLicense publishLicense, CryptoProvider cryptoProvider);
        public static EncryptedPackageEnvelope CreateFromPackage(string envelopeFileName, Stream packageStream, PublishLicense publishLicense, CryptoProvider cryptoProvider);
        public void Dispose();
        protected virtual void Dispose(bool disposing);
        public void Flush();
        public Package GetPackage();
        public static bool IsEncryptedPackageEnvelope(Stream stream);
        public static bool IsEncryptedPackageEnvelope(string fileName);
        public static EncryptedPackageEnvelope Open(Stream envelopeStream);
        public static EncryptedPackageEnvelope Open(string envelopeFileName);
        public static EncryptedPackageEnvelope Open(string envelopeFileName, FileAccess access);
        public static EncryptedPackageEnvelope Open(string envelopeFileName, FileAccess access, FileShare sharing);
    }
    public delegate void InvalidSignatureEventHandler(object sender, SignatureVerificationEventArgs e);
    public abstract class Package : IDisposable {
        // Existing type, below are added members
        protected Package(FileAccess openFileAccess, bool streaming);
    }
    public class PackageDigitalSignature {
        public CertificateEmbeddingOption CertificateEmbeddingOption { get; }
        public Signature Signature { get; set; }
        public PackagePart SignaturePart { get; }
        public string SignatureType { get; }
        public byte[] SignatureValue { get; }
        public ReadOnlyCollection<Uri> SignedParts { get; }
        public ReadOnlyCollection<PackageRelationshipSelector> SignedRelationshipSelectors { get; }
        public X509Certificate Signer { get; }
        public DateTime SigningTime { get; }
        public string TimeFormat { get; }
        public List<string> GetPartTransformList(Uri partName);
        public VerifyResult Verify();
        public VerifyResult Verify(X509Certificate signingCertificate);
    }
    public sealed class PackageDigitalSignatureManager {
        public PackageDigitalSignatureManager(Package package);
        public CertificateEmbeddingOption CertificateOption { get; set; }
        public static string DefaultHashAlgorithm { get; }
        public string HashAlgorithm { get; set; }
        public bool IsSigned { get; }
        public IntPtr ParentWindow { get; set; }
        public Uri SignatureOrigin { get; }
        public static string SignatureOriginRelationshipType { get; }
        public ReadOnlyCollection<PackageDigitalSignature> Signatures { get; }
        public string TimeFormat { get; set; }
        public Dictionary<string, string> TransformMapping { get; }
        public event InvalidSignatureEventHandler InvalidSignatureEvent;
        public PackageDigitalSignature Countersign();
        public PackageDigitalSignature Countersign(X509Certificate certificate);
        public PackageDigitalSignature Countersign(X509Certificate certificate, IEnumerable<Uri> signatures);
        public PackageDigitalSignature GetSignature(Uri signatureUri);
        public void RemoveAllSignatures();
        public void RemoveSignature(Uri signatureUri);
        public PackageDigitalSignature Sign(IEnumerable<Uri> parts);
        public PackageDigitalSignature Sign(IEnumerable<Uri> parts, X509Certificate certificate);
        public PackageDigitalSignature Sign(IEnumerable<Uri> parts, X509Certificate certificate, IEnumerable<PackageRelationshipSelector> relationshipSelectors);
        public PackageDigitalSignature Sign(IEnumerable<Uri> parts, X509Certificate certificate, IEnumerable<PackageRelationshipSelector> relationshipSelectors, string signatureId);
        public PackageDigitalSignature Sign(IEnumerable<Uri> parts, X509Certificate certificate, IEnumerable<PackageRelationshipSelector> relationshipSelectors, string signatureId, IEnumerable<DataObject> signatureObjects, IEnumerable<Reference> objectReferences);
        public static X509ChainStatusFlags VerifyCertificate(X509Certificate certificate);
        public VerifyResult VerifySignatures(bool exitOnFailure);
    }
    public static class PackUriHelper {
       // Existing type, below are added members
        public static int ComparePackUri(Uri firstPackUri, Uri secondPackUri);
        public static Uri Create(Uri packageUri);
        public static Uri Create(Uri packageUri, Uri partUri);
        public static Uri Create(Uri packageUri, Uri partUri, string fragment);
        public static Uri GetPackageUri(Uri packUri);
        public static Uri GetPartUri(Uri packUri);
    }
    public class RightsManagementInformation {
        public CryptoProvider CryptoProvider { get; set; }
        public void DeleteUseLicense(ContentUser userKey);
        public IDictionary<ContentUser, UseLicense> GetEmbeddedUseLicenses();
        public PublishLicense LoadPublishLicense();
        public UseLicense LoadUseLicense(ContentUser userKey);
        public void SavePublishLicense(PublishLicense publishLicense);
        public void SaveUseLicense(ContentUser userKey, UseLicense useLicense);
    }
    public class SignatureVerificationEventArgs : EventArgs {
        public PackageDigitalSignature Signature { get; }
        public VerifyResult VerifyResult { get; }
    }
    public class StorageInfo {
        public string Name { get; }
        public StreamInfo CreateStream(string name);
        public StreamInfo CreateStream(string name, CompressionOption compressionOption, EncryptionOption encryptionOption);
        public StorageInfo CreateSubStorage(string name);
        public void DeleteStream(string name);
        public void DeleteSubStorage(string name);
        public StreamInfo GetStreamInfo(string name);
        public StreamInfo[] GetStreams();
        public StorageInfo GetSubStorageInfo(string name);
        public StorageInfo[] GetSubStorages();
        public bool StreamExists(string name);
        public bool SubStorageExists(string name);
    }
    public class StreamInfo {
        public CompressionOption CompressionOption { get; }
        public EncryptionOption EncryptionOption { get; }
        public string Name { get; }
        public Stream GetStream();
        public Stream GetStream(FileMode mode);
        public Stream GetStream(FileMode mode, FileAccess access);
    }
    public enum VerifyResult {
        CertificateRequired = 2,
        InvalidCertificate = 3,
        InvalidSignature = 1,
        NotSigned = 5,
        ReferenceNotFound = 4,
        Success = 0,
    }
    public sealed class ZipPackagePart : PackagePart {
        protected override Stream GetStreamCore(FileMode modestreamFileMode, FileAccess accessstreamFileAccess);
    }
}
```

Here's the source: https://referencesource.microsoft.com/#WindowsBase/Base/System/IO/Packaging/EncryptedPackage.cs

To unblock myself I ported that codebase to .NETCore, but didn't attempt to reconcile it with System.IO.Packaging.

**Comment by danmosemsft** [comment]
@ericstj it seems this is WPF owned. Do you have a github ID for them?

**Comment by ericstj** [comment]
@vatsan-madhavan-msft and @rladuca

**Comment by safern** [comment]
Can I take this one or does it need further discussion?

**Comment by ericstj** [comment]
@onovotny please open a separate issue for that.  Let's let this issue track reconciling the WPF System.IO.Packaging API and behavior with that which already exists in CoreFx.

**Comment by ericstj** [comment]
@stevenbrix has done this reconciliation.  Some API has been added back to System.IO.Packaging.dll in corefx, the rest remains in WindowsBase in WPF.
